table.cc

```diff
@@ -2344,3 +2346,39 @@ table::as_mutation_source_excluding(std::vector<sstables::shared_sstable>& ssts)
         return this->make_reader_excluding_sstables(std::move(s), std::move(permit), ssts, range, slice, pc, std::move(trace_state), fwd, fwd_mr);
     });
 }
+
+future<> table::update_repair_time(dht::token_range range, gc_clock::time_point repair_time) {
+    _repair_time_map += std::make_pair(locator::token_metadata::range_to_interval(range), repair_time);
+    return make_ready_future<>();
+}
+
+gc_clock::time_point
+table::get_repair_based_tombstone_gc_before(const dht::decorated_key& dk) const {
+    // FIXME: Make propagation_delay tunable
+    constexpr auto propagation_delay = std::chrono::seconds(3600);
+    if (!use_repair_based_tombstone_gc()) {
+        auto repair_timestamp = gc_clock::now();
+        auto repair_based_tombstone_gc_before = saturating_subtract(repair_timestamp, propagation_delay);
+        tlogger.trace("Get repair based gc time for ks={}, table={}, dk={}, repair_timestamp={}, propagation_delay={}, repair_based_tombstone_gc_before={}, do not use repair_based_tombstone_gc",
+                schema()->ks_name(), schema()->cf_name(), dk, repair_timestamp, propagation_delay.count(), repair_based_tombstone_gc_before);
+        return repair_based_tombstone_gc_before;
+    }
+    const auto it = _repair_time_map.find(dk.token());
+    if (it == _repair_time_map.end()) {
+        return gc_clock::time_point::min();
+    } else {
+        auto& repair_timestamp = it->second;
+        auto repair_based_tombstone_gc_before = saturating_subtract(repair_timestamp, propagation_delay);
+        tlogger.trace("Get repair based gc time for ks={}, table={}, dk={}, repair_timestamp={}, propagation_delay={}, repair_based_tombstone_gc_before={}",
+                schema()->ks_name(), schema()->cf_name(), dk, repair_timestamp, propagation_delay.count(), repair_based_tombstone_gc_before);
+        return repair_based_tombstone_gc_before;
```

Comment:
I think we'll need some caching here (for a range) to avoid slowness when compacting sstables with lots of tombstones.

Maybe the function can return a { token_range, timestamp } so we can avoid the function call altogether if we hit the same range again.

Reply:
I will think about it. Since it is an optimization and this series is very big. I would propose to do the cache optimization as a follow up after the main part is merged.

Reply:
The series is quite small for what it delivers. I agree to do it in a follow-up if you measure the effect on compaction.

Reply:
ok